command-line option to control unit of time

ta946 opened the pull request:
https://github.com/rkern/line_profiler/issues/136#issue-387949622
add a command line option -u --unit to specify the unit of time displayed by line_profiler for ease of use and ease of reading the result.
Windows defaults to 1e-7 whereas specifying 1e-6 would be easier to read and mentally parse.
The user can easily change the unit if the runtimes are expected to be long eg: seconds

eg: to specify a time unit of 1us on the output of kernprof "script.lprof"
python -m line_profiler ./script.lprof -u 1e-6

Erotemic commented:
While I am one of the new maintainers of `line_profiler`, I'm not the original author, and I don't have full knowledge of the system, so bear with me. 

It looks like the existing repo has some ability to handle an `output_unit`, so all that you're doing here is exposing it via the command line utility? If so this seems like an easy accept. 

I do have two requests though: 

1. Expand the help text on the `unit` option to describe that the unit of time should be a float and that it will be interpreted. Maybe give an example saying specifying 1 will return output in seconds and 1e-3 will return milliseconds. (Is there any use in allowing the user to specify string codes like: `s`, `ms`, `us`, `ns`? Would this just add unnecessary complexity?)

2. Can you bump the `__version__` attribute in `line_profiler.py` to `3.1.2` and start an appropriate section in the `CHANGELOG.md` document?

ta946 commented:
Yes, this PR is just exposing the output_unit, that the repo already uses, to the command line.

1. there are quite a few packages that can parse the unit but i dont like the idea of adding a dependancy just for that.
Doing it manually would be easy if people just stuck to using a single letter for unit (s, ms, us), but if they start using words (milli, micro, etc.) then it would be alot more of a headache to parse by trying to accomodate for every possible input.
It also opens it up to failure due to user error eg. if an incorrect letter or word is used.

2. a "command line options" section for line_profiler might also need to be added in the readme. I wouldn't know the appropriate format and wording, so im afraid ill have to leave that part to you

Erotemic commented:
I think just keeping the floating point format is reasonable. It makes more sense now that there is more info in the help. I also think the help is sufficient, so no need to modify the README. 

It looks like there is some 3.5 issue causing the dashboard to fail. There is some version issue with pytest and pytest-cov. I spent a little time looking into it, but was unable to replicate it on my machine and clearing the travis cache didn't seem to work. I'd like to get that fixed before I merge this, but other than that this LGTM.

ta946 commented:
regarding the readme, im suggested adding a command line section because atm there are no command line arguments for line_profiler and there is no indication that this option would be available for the user unless they brought up the help for line_profiler

artemisart commented:
@Erotemic  I looked at the travis log and pytest cov requires pytest>=4.6 but 4.1 is installed, could change the file `requirements/tests.txt` to put `pytest>=4.6`, or maybe just be safe and use `pytest>=6.1` ? (I think I cannot add commits to an existing PR as I'm not the author).

Erotemic commented:
@ta946 Please rebase this branch on master. Thank's to @artemisart's PR it should now pass the CI and be mergeable.

ta946 commented:
@Erotemic not sure if I rebased the branch from my fork correctly?
I rebased the branch onto "upstream/master" and it asked me to pull before I can push which ended up with a merge after the rebase

artemisart commented:
The patch on requirements/tests.txt should'nt appear here but it's not a big issue.

ta946 commented:
yea looks like its comparing from when the fork was created. i thought using a PR from a fork would be the cleanest but it seems to be causing more of a mess. Its my first time trying to use githubs PRs so I worry ive done something wrong.
I wouldnt mind closing my PRs and either someone else or I create new PRs if thats less of a headache

Erotemic commented:
@ta946 either check the "allow maintainers to edit" checkbox or do this: 

* checkout master and pull from this repo: `git checkout master; git pull origin master`

* rebase on master `git rebase -i master`, then mark that all but the first commits should be squashed (by replacing the "pick" with an "s") . There will be one conflict that you'll have to fix, but after you do that then force push back to this branch. Then it will be in a clean state. 

I've also pushed up a rebased version of this branch to this remote. So, the last alternative thing you could do is to `git fetch origin` (assuming origin is pointing to this remote) and then `git reset origin/commandline-time-unit-rebase --hard` on your branch and then `git push ta946 --force` (assuming your remote is called `ta946`).

artemisart commented:
@Erotemic I integrated these changes in #34

Erotemic commented:
Closed in favor of #34
